src/main/scala/components/hazardmemstall.scala

```diff
@@ -0,0 +1,76 @@
+// This file contains the hazard detection unit
+
+package dinocpu
+
+import chisel3._
+
+/**
+ * The hazard detection unit, modified for memory stalling
+ *
+ * Input:  rs1, the first source register number
+ * Input:  rs2, the first source register number
+ * Input:  idex_memread, true if the instruction in the ID/EX register is going to read from memory
+ * Input:  idex_rd, the register number of the destination register for the instruction in the ID/EX register
+ * Input:  exmem_taken, if true, then we are using the nextpc in the EX/MEM register, *not* pc+4.
+ * Input:  imem_good, high if instruction memory is idle and ready.
+ *         If either imem_good or dmem_good are false then CPU will stall
+ * Input:  dmem_good, high if data memory is idle and ready.
+ *
+ * Output: pcwrite, the value to write to the pc. If 0, pc+4, if 1 the next_pc from the memory stage,
+ *         if 2, then the last pc value (2 stalls the pipeline)
+ * Output: ifid_bubble, if true, we should instet a bubble in the IF/ID stage
+ * Output: idex_bubble, if true, we should insert a bubble in the ID/EX stage
+ * Output: exmem_bubble, if true, we should insert a bubble in the EX/MEM stage
+ * Output: ifid_flush, if true, set the IF/ID register to 0
+ *
+ * For more information, see Section 4.7 and beginning of 4.8 of Patterson and Hennessy
+ * This follows the "Data hazards and stalls" section and the "Assume branch not taken" section
+ */
+class HazardUnitMemStall extends Module {
+  val io = IO(new Bundle {
+    val rs1          = Input(UInt(5.W))
+    val rs2          = Input(UInt(5.W))
+    val idex_memread = Input(Bool())
+    val idex_rd      = Input(UInt(5.W))
+    val exmem_taken  = Input(Bool())
+
+    val imem_good    = Input(Bool())
+    val dmem_good    = Input(Bool())
+
+    val pcwrite      = Output(UInt(2.W))
+    val ifid_bubble  = Output(Bool())
+    val idex_bubble  = Output(Bool())
+    val exmem_bubble = Output(Bool())
+    val ifid_flush   = Output(Bool())
+  })
+
+  // default
+  io.pcwrite      := 0.U
+  io.ifid_bubble  := false.B
+  io.idex_bubble  := false.B
+  io.exmem_bubble := false.B
+  io.ifid_flush   := false.B
+
+  // Load to use hazard.
+  when (io.idex_memread &&
+        (io.idex_rd === io.rs1 || io.idex_rd === io.rs2)) {
+    io.pcwrite     := 2.U
+    io.ifid_bubble := true.B
+    io.idex_bubble := true.B
+  }
+
+  // branch flush
+  when (io.exmem_taken) {
+    io.pcwrite := 1.U // use the PC from mem stage
+    io.ifid_flush  := true.B
+    io.idex_bubble  := true.B
+    io.exmem_bubble := true.B
+  }
+
+  // mem stall
+  when (~(io.imem_good && io.dmem_good)) {
+    io.pcwrite     := 2.U // freeze PC
+    io.idex_bubble  := true.B
+    //io.exmem_bubble := true.B
+  }
+}
```

Comment:
I noted the comment about my changes not having newlines at the end of files. My IDE is currently configured to add newlines right now and indeed my current copy of the repository has them. Maybe Git is stripping away the newline? I would have to check.